### PR TITLE
Galilean PSATD with shift  

### PR DESCRIPTION
**Pull request by oshapoval**
**Description edited by @RemiLehe **

# Overview

This implements the Galilean algorithm, as described in [this paper](https://aip.scitation.org/doi/full/10.1063/1.4964770) and [this paper](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.94.053305).

The Galilean algorithm is a spectral algorithm that suppresses the Cherenkov instability for relativistic plasma flows. 

# Implementation details

- The Galilean algorithm is implemented as a sub-class of `SpectralBaseAlgorithm`. This allows to modify only the parts of the code that do differ from the standard PSATD algorithm (field pusher in k space and initialization of the associated coefficients), while reusing the rest of the code that was already implemented for the standard PSATD algorithm (i.e. fourier transforms back and forth)

- One key part in this PR is keeping track of the Galilean shift. More specifically, in the Galilean algorithm, the grid points are assumed to move with a given velocity. However, in `amrex` gridpoints are typically static. We can move them by using `ResetProbDomain`, but it is problematic to do so within one PIC iterations. Therefore, here we took the following approach:
   * Use `ResetProbDomain` only at the end of each PIC iteration to shift the positions of the grid points (just before calling the moving window)
    * Within one PIC iteration: calculate the shift due to the galilean velocity, since the **last time that we called `ResetProbDomain**. This is done everytime that a particle-grid operation is performed (e.g. deposition, gathering). In order to be more confident that we found all of these places, we relied on the fact that a particle-grid operation typically calls the function `LowerCorner`, and we modified this function so that it *has* to take a galilean shift as argument.

### CODE REVIEW

**Comment by MaxThevenet** [comment]
Thanks for this PR @oshapoval!
Could you target `master` instead of `dev` for this PR? 
Also, could you add some more details into the PR description? An example can be seen at https://github.com/ECP-WarpX/WarpX/pull/616

Thanks!

File: Source/Utils/WarpXMovingWindow.cpp [hunk 1]
```diff
@@ -348,6 +348,43 @@ WarpX::shiftMF (MultiFab& mf, const Geometry& geom, int num_shift, int dir,
     }
 }
 
+void
+WarpX::ShiftGalileanBoundary ()
+{
+    Real cur_time = t_new[0];
+    WarpX:: time_of_last_gal_shift;
```
Comment (RemiLehe) [inline Source/Utils/WarpXMovingWindow.cpp]: The above statement does not seem to do anything. Could you remove it?

File: Source/Particles/Pusher/UpdatePosition.H [hunk 2]
```diff
@@ -32,4 +32,28 @@ void UpdatePosition(amrex::ParticleReal& x, amrex::ParticleReal& y, amrex::Parti
     z += uz * inv_gamma * dt;
 }
 
+/* \brief Push the particle's positions over one timestep,
+ *    given the value of its momenta `ux`, `uy`, `uz` */
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void UpdatePositionGalilean(
```
[the remark this answers is not in the document]
Comment (oshapoval) [reply Source/Particles/Pusher/UpdatePosition.H]: Removed temporary used particle's push for Galilean PSATD. Done in commit 7fe30c8e4ab.

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 3]
```diff
@@ -53,6 +54,11 @@ void doDepositionShapeN(const GetParticlePosition& GetPosition,
     // Whether ion_lev is a null pointer (do_ionization=0) or a real pointer
     // (do_ionization=1)
     const bool do_ionization = ion_lev;
+
```
[the remark this answers is not in the document]
Comment (oshapoval) [reply Source/Particles/Deposition/CurrentDeposition.H]: Removed temporary used galilean shift. Done in commit d8cdd63eb9f3.

File: Source/Utils/WarpXMovingWindow.cpp (hunk 1 above)
Comment (oshapoval) [reply Source/Utils/WarpXMovingWindow.cpp]: Done in commit 3d584488063.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+#include <SpectralBaseAlgorithm.H>
+
+/* \brief Class that updates the field in spectral space
+ * and stores the coefficients of the corresponding update equation.
+ */
+class GalileanAlgorithm : public SpectralBaseAlgorithm
+{
+    public:
+        GalileanAlgorithm(const SpectralKSpace& spectral_kspace,
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H]: please add a Doxygen string that describes the parameters. [Here is a little example](https://github.com/openPMD/openPMD-api/blob/a545854367624984555a1d2a06463e88796df3fd/include/openPMD/auxiliary/Date.hpp#L30-L35).

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+                         const amrex::DistributionMapping& dm,
+                         const int norder_x, const int norder_y,
+                         const int norder_z, const bool nodal,
+                         const amrex::Array<amrex::Real,3>& v_galilean,
+                         const amrex::Real dt);
+        // Redefine update equation from base class
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H]: please add a Doxygen string that describes the parameters of `pushSpectralFields`.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        virtual void pushSpectralFields(SpectralFieldData& f) const override final;
+        virtual int getRequiredNumberOfFields() const override final {
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H]: please add a Doxygen string that describes the parameters and return value of `getRequiredNumberOfFields`.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            return SpectralFieldIndex::n_fields;
+        };
+        void InitializeSpectralCoefficients(const SpectralKSpace& spectral_kspace,
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.H]: please add a Doxygen string that describes the parameters and side effects of `InitializeSpectralCoefficients`.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.cpp [hunk 8]
```diff
@@ -0,0 +1,239 @@
+#include <GalileanAlgorithm.H>
+#include <WarpXConst.H>
+#include <cmath>
+
+using namespace amrex;
+
+/* \brief Initialize coefficients for the update equation */
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanAlgorithm.cpp]: this only goes in the header `.H` file :)

**Pull request merged**